service/state/core.go

```diff
@@ -68,50 +67,24 @@ func NewCore(dir string) (*Core, error) {
 
 	var height uint64
 	var commit flow.StateCommitment
-	err = db.View(func(tx *badger.Txn) error {
-
-		// first we get the last commit
-		if err := storage.RetrieveLastCommit(&commit)(tx); err != nil {
-			return fmt.Errorf("could not retrieve last commit: %w", err)
-		}
-
-		// then we get the height associated with it
-		if err := storage.RetrieveHeightByCommit(commit, &height)(tx); err != nil {
-			return fmt.Errorf("could not retrieve height from commit: %w", err)
-		}
-		return nil
+	trie, err := trie.NewEmptyMTrie(pathfinder.PathByteSize)
+	if err != nil {
+		return nil, fmt.Errorf("could not initialize empty trie: %w", err)
+	}
+
+	empty := trie.RootHash()
+	err = db.Update(func(txn *badger.Txn) error {
+		return storage.Fallback(
+			storage.Combine(storage.RetrieveLastCommit(&commit), storage.RetrieveHeightByCommit(commit, &height)),
+			storage.Combine(storage.SaveLastCommit(empty), storage.SaveHeightForCommit(0, empty)),
+		)(txn)
 	})
-	if err != nil && !errors.Is(err, badger.ErrKeyNotFound) {
-		return nil, fmt.Errorf("could not retrieve last commit: %w", err)
+	if err != nil {
+		return nil, fmt.Errorf("could not retrieve last height and commit: %w", err)
 	}
-	if errors.Is(err, badger.ErrKeyNotFound) {
-
-		// create an empty trie root hash as last commit
-		trie, err := trie.NewEmptyMTrie(pathfinder.PathByteSize)
-		if err != nil {
-			return nil, fmt.Errorf("could not initialize empty trie: %w", err)
-		}
-		commit = trie.RootHash()
-
-		err = db.Update(func(tx *badger.Txn) error {
-
-			// store the empty root hash as last commit
-			err = storage.SaveLastCommit(commit)(tx)
-			if err != nil {
-				return fmt.Errorf("could not persist last commit: %w", err)
-			}
-
-			// map the last commit to zero height
-			err = storage.SaveCommitForHeight(commit, height)(tx)
-			if err != nil {
-				return fmt.Errorf("could not persist commit index: %w", err)
-			}
-
-			return nil
-		})
-		if err != nil {
-			return nil, fmt.Errorf("could not bootstrap last commit: %w", err)
-		}
+
+	if len(commit) == 0 {
+		commit = empty
 	}
```

Comment:
We can simply not use `empty`, and use `commit` everywhere. Then this if clause becomes unnecessary.